src/index.js

```diff
@@ -19,60 +19,112 @@ const commands = Object.keys(queries).map(queryName => {
       const queryImpl = queries[queryName]
       const baseCommandImpl = doc => {
         const container = getContainer(waitOptions.container || doc)
-        return waitForElement(() => queryImpl(container, ...args), {
-          ...waitOptions,
-          container,
-        })
+        return queryImpl(container, ...args)
       }
-      let commandImpl
-      if (
-        queryName.startsWith('queryBy') ||
-        queryName.startsWith('queryAllBy')
-      ) {
-        commandImpl = doc =>
-          baseCommandImpl(doc).catch(_ =>
-            doc.querySelector('.___cypressNotExistingSelector'),
-          )
-      } else {
-        commandImpl = doc => baseCommandImpl(doc)
+      let commandImpl = doc => baseCommandImpl(doc)
+
+      var inputArr = args.filter(filterInputs);
+
+      var consoleProps = {
+        // TODO: Would be good to completely separate out the types of input into their own properties
+        input: inputArr
       }
-      const thenHandler = new Function(
-        'commandImpl',
-        `
-            return function Command__${queryName}(thenArgs) {
-              return commandImpl(thenArgs.document)
-            }
-          `,
-      )(commandImpl)
+
+      Cypress.log({
+        $el: inputArr,
+        name: queryName,
+        message: inputArr,
+        consoleProps: () => consoleProps
+      });
+
       return cy
         .window({log: false})
-        .then({timeout: waitOptions.timeout + 100}, thenHandler)
-        .then(subject => {
-          Cypress.log({
-            $el: subject,
-            name: queryName,
-            message: args.filter(value => {
-              if (Array.isArray(value) && value.length === 0) {
-                return false
-              }
-              if (value instanceof RegExp) {
-                return value.toString()
+        .then((thenArgs) => {
+          const getValue = () => {
+            var result;
+            try {
+              var value = commandImpl(thenArgs.document);
+              result = Cypress.$(value);
+            }
+            catch (err) {
+              // Catch exceptions where it can't find the element, so we can retry
+              if (/Unable to find an element with the text/.test(err.message))
```

Comment:
I was originally catching any exception to getBy working as expected, when testing other functionality I realised that the retry-ability stuff would try everything even when the failure is correct. For example getBy, findBy and queryBy all throw an error when matching multiple results, and that scenario is less likely something the user wants to retry, once it fails that is probably the correct result.

If there are other error messages for getByLabelText, getByDisplayValue, getByTestId, getByRole etc, then yes we definitely need to add those or create a specific error type in dom-testing-library and catch those.

I'll defer the decision to someone with greater knowledge of dom-testing-library for that one.

Reply:
I've since seen that getByLabelText doesn't retry, due to the error message not being as you mention. We can add the relevant messages for each of those queries, but I wonder if there is a way to catch all of those in a more succinct and deliberate way?

@kentcdodds can we create a specific type of error message for get* exceptions in dom-testing-library? That way we can test on something more concrete than the error message text, which could change over time.

Reply:
Something along those lines would be fine yes. What does DTL stand for?

Reply:
The main thing we need to know is how many elements have been found. If it’s 0 then we know we need to re-run the function, if it’s more than that then we let it throw.

Either adding a property to that affect on said errors would work, whether you want to use a new class of error is up to you.